### PR TITLE
Validate function

### PR DESCRIPTION
**Pull request by MayElnathan**

### CODE REVIEW

File: ReactApp/src/utils/rest-api-call.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import Cookies from 'js-cookie';
+import {validateParameters} from '../../../src/main/validateParameters'
```
Comment (RoiTief) [inline ReactApp/src/utils/rest-api-call.js:2]: does this work even though this file is completely out of scope?

File: src/main/validateParameters.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+      ...expectedObj,
+      callingUser: {
+        username: PRIMITIVE_TYPES.STRING,
+        type: PRIMITIVE_TYPES.STRING,
+      },
+    };
+  }
+
+  for (const key in expectedObj) {
+    if (expectedObj.hasOwnProperty(key)) {
```
Comment (RoiTief) [inline src/main/validateParameters.js:84]: isn't that always try by definition of the for loop?

File: src/main/validateParameters.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+      // Check if the key exists in realObj
+      if (!realObj.hasOwnProperty(key)) {
+        throw new EMSError(ERROR_MSGS.MISSING_KEY(key,callingFunctionName), ERROR_CODES.MISSING_KEY);
+      }
+
+      // Get the expected type
+      const expectedType = expectedObj[key];
+      const realValue = realObj[key];
+
+      if (prohibitNull && realValue === null) {
```
Comment (RoiTief) [inline src/main/validateParameters.js:94]: What if some input is mandatory and some optional?
would I need to call this function twice?

File: src/main/validateParameters.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        throw new EMSError(ERROR_MSGS.NULL_VALUE(key,callingFunctionName), ERROR_CODES.NULL_VALUE);
+      }
+
+      // Check the type
+      
+      // Primitive type check
+      if (Object.values(PRIMITIVE_TYPES).includes(expectedType)) {
```
Comment (RoiTief) [inline src/main/validateParameters.js:101]: seeing you using this `includes` method teaches me you don't review my code properly. 
In another PR of mine I recreated this rather awkwardly.

File: src/main/validateParameters.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+      } 
+        // Class type check
+        else if (typeof expectedType === 'function') {
+        if (!(realValue instanceof expectedType)) {
+          throw new EMSError(ERROR_MSGS.TYPE_MISMATCH(key, expectedType.name, realValue.constructor.name,callingFunctionName), ERROR_CODES.TYPE_MISMATCH);
+        }
+      }
+        // Nested object type check
+        else if (typeof expectedType === 'object') {
+          validateParameters(realValue, expectedType, prohibitNull, false);
```
Comment (RoiTief) [inline src/main/validateParameters.js:114]: beautiful recursion :)

File: src/main/ErrorMessages.js [hunk 6]
```diff
@@ -38,6 +38,14 @@ const SESSION_PROCESS_ERROR_MSGS = {
     SESSION_NOT_IN_USE: 'Session is not in use',
 }
 
+const FUNCTION_PARAMETERS_ERROR_MSGS = {
+    MISSING_KEY: (key,callingFunctionName) => `Missing key: ${key}. ${callingFunctionName}`,
```
Comment (RoiTief) [inline src/main/ErrorMessages.js:42]:
```suggestion
    MISSING_KEY: (key,callingFunctionName) => `Missing key: '${key}' when using '${callingFunctionName}'`,
```
Fix other messages in the same way

File: ReactApp/src/utils/rest-api-call.js (hunk 1 above)
Comment (MayElnathan) [reply ReactApp/src/utils/rest-api-call.js:2]: Yes, seems like it ,as far as i have tested...

File: src/main/ErrorMessages.js (hunk 6 above)
Comment (MayElnathan) [reply src/main/ErrorMessages.js:42]: callingFunctionName do it automatically, changing for the key

File: src/main/validateParameters.js (hunk 2 above)
Comment (MayElnathan) [reply src/main/validateParameters.js:84]: yep, will be removed

File: src/main/validateParameters.js (hunk 3 above)
Comment (MayElnathan) [reply src/main/validateParameters.js:94]: Actually yes.
I hope we wont get to this at all, it will be a total mess, and super hard to maintain if we allow only some fields to be null and the other not.

I made it so we can choose (with null or without null, for all fields) and change it easily in the future, but it will be really bad if we will do a mix

File: src/main/validateParameters.js (hunk 4 above)
Comment (MayElnathan) [reply src/main/validateParameters.js:101]: Maybe, I sorry if I did, I try my best....
Unfortunately I sometimes do a CR after a work day and 2 classes at 0:00, not that it is a good excuse, but I can see how it can happens

**Review by OfekN11 — CHANGES_REQUESTED** [review]
Add array support
